### PR TITLE
[Abandoned] Add StarBorder and StarBorder.polygon

### LINKED ISSUE #107442: [Request] Scallop / Sunflower / Star Material 3 Shape
I think it would be nice for Flutter to adopt the Clover / Scallop / Sunflower shape from Android 12 / M3.

It could be a custom OutlinedBorder with custom math like this: https://stackoverflow.com/a/25840319
You probably have access to the Android 12 source code, so that could also be helpful.

Figma uses three pieces of information:

- pointCount (number of spikes, 5 for a star)
- innerRadius (%, if it should grow inside or outside)
- cornerRadius

This is a widget that feels heavy to use, but when you need it is really nice to have. I would love to see in the framework.

image

Frame 4

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This adds `StarBorder`, along with the interpolations to/from `CircleBorder`, `RoundedRectBorder`, and `StadiumBorder`.

The example looks like [this Dartpad](https://dartpad.dev/?id=28c825b7b91963c11c17cc4b6a0c1324). (If it won't load because GitHub's API is overloaded, copy the Gist [here](https://gist.github.com/gspencergoog/28c825b7b91963c11c17cc4b6a0c1324) into your own [DartPad](https://dartpad.dev)).

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/107442

## Tests
 - Needs tests (hence the Draft).

### CODE REVIEW

**Comment by bernaferrari** [comment]
Wow. That's so good. I would never be able to do something like this that fast.

**Comment by bernaferrari** [comment]
Did you take a look at the Android implementation? Or is it 100% from your mind? It is so sophisticated...

**Comment by bernaferrari** [comment]
This is small, don't worry if it is too hard to fix, but lerp 0.00 to 0.01 changes the size:

https://user-images.githubusercontent.com/351125/180669315-516a8781-536a-40d3-b193-b3102c886f5a.mov

.

**Comment by bernaferrari** [comment]
Suggestion: add a constructor that calculates `valleyRounding` and `pointRounding` from `side.borderRadius`. Could be something like `clampDouble(borderRadius / rect, 0, 0.5)` and then it applies this to both `valleyRounding` and `pointRounding`. You would hide two (kind of) confusing variables from the user and make star work with the BorderRadius user already know and love.

Also, I don't know what I did here, but it doesn't seem fine (again, if it is too hard or just an edge case, don't worry):
image

**Comment by gspencergoog** [comment]
What you're seeing there in the last one you posted is numerical instability as the inner radius ratio goes to zero.  I'm now clamping it to a minimum of 1e-6, which should avoid that.

As for your suggestion around `borderRadius`, I'm not sure that it's any more clear.  I can't calculate the exact radius because I'm scaling it all at the end to keep it the same size as the inner radius ratio goes to zero (which is also the source of the numerical instability), so the number in the radius would be pretty arbitrary (e.g. you'd specify 100, but get back something with radius 10, and as you change the inner radius ratio you'd have to also change the `borderRadius` to keep it drawn at a constant size). Also, the squash parameter would change any Radius.circular to an ellipse.

**Comment by bernaferrari** [comment]
> `borderRadius`

Hmmm.. :( makes sense. I still wish there were a 'simpler' way of dealing with `valleyRounding` and `pointRounding`.

File: examples/api/lib/painting/star_border/star_border.0.dart [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+  @override
+  void initState() {
+    super.initState();
+    _model.addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    _model.removeListener(_modelChanged);
```
Comment (HansMuller) [inline examples/api/lib/painting/star_border/star_border.0.dart:50]: textController.dispose()

File: examples/api/lib/painting/star_border/star_border.0.dart [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+  Widget build(BuildContext context) {
+    return DefaultTextStyle(
+      style: const TextStyle(
+        color: Colors.black,
+        fontSize: 14.0,
+        fontFamily: 'Roboto',
+        fontStyle: FontStyle.normal,
+      ),
+      child: SliderTheme(
+        data: SliderTheme.of(context).copyWith(),
```
Comment (HansMuller) [inline examples/api/lib/painting/star_border/star_border.0.dart:67]: This seems odd. If it's necessary, a comment that explains why would help.

File: examples/api/lib/painting/star_border/star_border.0.dart [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+              color: Colors.grey.shade200,
+              child: Options(_model),
+            ),
+            Padding(
+              padding: const EdgeInsets.all(18.0),
+              child: Row(
+                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+                children: <Widget>[
+                  Expanded(
+                    child: Container(
```
Comment (HansMuller) [inline examples/api/lib/painting/star_border/star_border.0.dart:80]: Could factor out a widget here and below, like:
```dart
class StarBorderView extends StatelessWidget {
  const StarBorderView(this.star, this.title);

  final StarBorder star;
  final String title;

  @override
  Widget build(BuildContext) {
    return Container(
      key: UniqueKey(),
      alignment: Alignment.center,
      padding: const EdgeInsets.all(20),
      width: 150,
      height: 100,
      decoration: ShapeDecoration(
        color: Colors.blue.shade100,
        shape: star,
      ),
      child: const Text(title),
    );
  }
}
```

There are other places where you could do this kind of thing. Not a big deal.

File: examples/api/lib/painting/star_border/star_border.0.dart [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+                    color: Colors.black12,
+                    margin: const EdgeInsets.all(16.0),
+                    padding: const EdgeInsets.all(16.0),
+                    child: SelectableText(_model.starCode),
+                  ),
+                ),
+                Expanded(
+                  child: Container(
+                    color: Colors.black12,
+                    margin: const EdgeInsets.all(16.0),
```
Comment (HansMuller) [inline examples/api/lib/painting/star_border/star_border.0.dart:138]: padding: const EdgeInsets.all(32) has the same effect?

File: examples/api/lib/painting/star_border/star_border.0.dart [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+
+  @override
+  void initState() {
+    super.initState();
+    _model.addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
```
Comment (HansMuller) [inline examples/api/lib/painting/star_border/star_border.0.dart:49]: Conventionally super.dispose() is last (here and elsewhere).

File: packages/flutter/lib/src/painting/star_border.dart [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+    double? valleyRounding,
+    double? rotation,
+    double? squash,
+  }) {
+    return StarBorder(
+      side: side ?? this.side,
+      points: points ?? this.points,
+      rotation: rotation ?? this.rotation,
+      innerRadiusRatio: innerRadiusRatio ?? this.innerRadiusRatio,
+      pointRounding: valleyRounding ?? this.pointRounding,
```
Comment (HansMuller) [inline packages/flutter/lib/src/painting/star_border.dart:379]: Looks like a typo here => valleyRounding

**Review by HansMuller — APPROVED** [review]
A star is born!

LGTM

**Comment by bernaferrari** [comment]
ContinuousRectangleBorder can't lerp to anything, it is one of the weirdest OutlinedBorders. Maybe you could apply what you learned in the star to make it better? Maybe it could extend StarBorder so it gets lerp for free?

It seems like your polygonal is pretty close to ContinuousRectangleBorder.

**Comment by gspencergoog** [comment]
> It seems like your polygonal is pretty close to ContinuousRectangleBorder and could help achieve what people want

Well, sort of.  I don't think I could generate a superellipse with the polygon generator: it would generate something similar but not quite the same, and because it would have to be rotated 45 degrees (and I use the non-rotated radius to calculate the scale), it would be smaller than they expected.

Best would be to just fix the `ContinuousRectangleBorder` to generate a squircles properly, and add some lerping targets to it. It's probably just not scaling them properly.

**Comment by bernaferrari** [comment]
> Best would be to just fix the ContinuousRectangleBorder

Shall I open the issue? lol
But I'm not sure I can help a lot, that's 100% path.

**Comment by gspencergoog** [comment]
I'm abandoning this PR for a fresh copy: https://github.com/flutter/flutter/pull/108489

This is because Skia Gold got really confused about which patchsets were which because of my force-pushing to this PR.

The contents are the same.

**Comment by gspencergoog** [comment]
Leaving this open for a few days so the Skia Gold team can take a look at it.

**Comment by Piinks** [comment]
@kjlubick @leandrolovisolo this is the PR associated with https://bugs.chromium.org/p/skia/issues/detail?id=13589
It is not clear what happened with Gold here.

**Pull request closed without merging**